### DISCUSSION TITLE
Load/cargo types cannot be mixed on same vehicle at the same time constrain

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#3112] [Routing (and legacy CP) questions] [answered]
Hi,

I used the search, but it returned nothing. Any assistance is appreciated.
We have a routing problem that requires that certain cargo types cannot be carried by the same vehicle at the same type. For example,
if I have cargo type A and cargo type B, a vehicle can carry any number of loads of type A or type B, but cannot carry type A and type B AT THE SAME TIME? How would I go define such a constraint?  Thank you.

**Comment by jmarca** [comment, marked as answer]
(I'm sure I've written this before in the forums.)

One approach is to use a dimension for each cargo type.

* dimension A:  1 if an A pickup, -1 if an A delivery;

* dimension B:  1 if an B pickup, -1 if an B delivery;

Then filter all of the nodes to eliminate possible next values
conditional on whether the A or B dimension is empty.

in not python-alike pseudo code:

first, A node pickups cannot go to B anything, and vice versa

```
for node in A_pickups:
   index = managerindexboilerplate
   for nextnode in B_pickup_or_deliver:
      nextindex = managerindexboilerplate
      routing.NextVar(index).RemoveValue(nextindex)
```

similarly, if node is an A delivery, it cannot go to a B delivery

```
for node in A_deliveries:
   index = managerindexboilerplate
   for nextnode in B_deliveries:
      nextindex = managerindexboilerplate
      routing.NextVar(index).RemoveValue(nextindex)
```

The tricky part is when a node is an A delivery, maybe the vehicle is
empty of all A.  If so, it can go to a B pickup.  The not so easy part
is to test if the vehicle is empty of A when it arrives at the B
pickup node

```
for node in  A_deliveries:
   index = manager.Index(node)...blah blah...
   for nextnode in B_pickups:
      nextindex  = manager.Index(nextnode)...blah blah...
      condition = A_dim.CumulVar(nextindex) > 0
      constraint = routing.NextVar(index) != nextindex
      expression = solver.ConditionalExpression(condition, constraint, 1)
      solver.Add(expression >= 1)
```

The difficulty is if you have lots and lots of different items that
can't be loaded at the same time, this will degrade the solver
performance, because every new dimension is a drain on the solver.

Hope that helps,

James



On Thu, Feb 03, 2022 at 07:37:06AM -0800, Oleg Dashevsky wrote:
> Hi,
>
> I used the search, but it returned nothing. Any assistance is appreciated.
> We have a routing problem that requires that certain cargo types cannot be carried by the same vehicle at the same type. For example,
> if I have cargo type A and cargo type B, a vehicle can carry any number of loads of type A or type B, but cannot carry type A and type B AT THE SAME TIME? How would I go define such a constraint?  Thank you.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3112
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by odashevsky** [reply]
Thank you very much. This gives me a very good idea what I need to do. Much appreciated.
